### ISSUE TITLE
Dashboard: Deleting a project cascades too many database queries

### ISSUE THREAD

**Issue by benjaoming** [readthedocs/readthedocs.org#10040]
Users cannot delete projects with a lot of data or views and end up with 502s.

On the Dashboard's "Delete Project" page, we use Django's built-in `DeleteView`:

https://github.com/readthedocs/readthedocs.org/blob/38fca6559bf954a448fec8786f78526e0a64a1e9/readthedocs/projects/views/private.py#L167-L180

It will call `.delete()` on the object to delete and then Django will call `.delete()` on any referencing FK that has `on_delete=models.CASCADE`

This likely becomes an issue because of FKs in models with a lot of data, such as this:

https://github.com/readthedocs/readthedocs.org/blob/38fca6559bf954a448fec8786f78526e0a64a1e9/readthedocs/analytics/models.py#L55-L63

The solution is likely to clean up some relations with a bulk delete query here:

https://github.com/readthedocs/readthedocs.org/blob/38fca6559bf954a448fec8786f78526e0a64a1e9/readthedocs/projects/models.py#L498-L504

**Comment by ericholscher** [comment]
I think we probably just want to delete things in a celery task, and not in process. To do that properly we need a bit more thought on UX, but in general even a 1-2 minute delete is better than a 502 that doesn't work :)

**Comment by benjaoming** [comment]
@ericholscher I like that the action is atomic and gives direct feedback. I don't think we need a celery task for this, we just need to _not_ hit the database once for every page visit that has been logged in analytics. We can probably wrap a db transaction around this.

As for celery tasks, the deletion of content and search indexes takes place async. Unfortunately, they are added BEFORE deleting the project, so the atomicity is a bit so-so.. would be best to delete this after the deletion of the project has gone well. The `post_delete` signal has access to a deferred object instance.

**Comment by humitos** [comment]
I'd propose to explore this idea:

* Use `on_delete=models.SET_NULL` for most of these relations where there are many objects involved
* Create a Celery tasks that performs a cleanup of these "orphaned" models that executes periodically and remove them

I'm assuming that `SET_NULL` is a lot faster than `CASCADE` since the later is "emulated" by Django (https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.CASCADE), but I'm might be wrong here.

@ericholscher 

> but in general even a 1-2 minute delete

If the query takes more than 30s, we will need to somehow exclude it from our general limitation.

@benjaoming 

> It will call `.delete()` on the object to delete and then Django will call `.delete()` on any referencing FK that has `on_delete=models.CASCADE`

This is not 100% true. Django _won't_ call `.delete()` on any referencing FK

> [Model.delete()](https://docs.djangoproject.com/en/4.1/ref/models/instances/#django.db.models.Model.delete) isn’t called on related models, but the [pre_delete](https://docs.djangoproject.com/en/4.1/ref/signals/#django.db.models.signals.pre_delete) and [post_delete](https://docs.djangoproject.com/en/4.1/ref/signals/#django.db.models.signals.post_delete) signals are sent for all deleted objects.

_(from https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.CASCADE)_

**Comment by stsewd** [comment]
> @ericholscher I like that the action is atomic and gives direct feedback. I don't think we need a celery task for this, we just need to not hit the database once for every page visit that has been logged in analytics. We can probably wrap a db transaction around this.

I'm +1 with this idea, just manually call `project.page_views.all().delete()` to do a bulk deletion before deleting the project itself.

**Comment by benjaoming** [comment]
> I'm assuming that SET_NULL is a lot faster than CASCADE since the later is "emulated" by Django (https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.CASCADE), but I'm might be wrong here.

@humitos I'm not sure that getting rid of a DB constraint here is a good idea since it opens up for a potentially unwanted state that we didn't design for. So then we'd have to go and check all the places where this FK is in use :mag:

**Comment by ericholscher** [comment]
> > @ericholscher I like that the action is atomic and gives direct feedback. I don't think we need a celery task for this, we just need to not hit the database once for every page visit that has been logged in analytics. We can probably wrap a db transaction around this.
> 
> I'm +1 with this idea, just manually call `project.page_views.all().delete()` to do a bulk deletion before deleting the project itself.

That seems like a good, easy first step. @benjaoming Want to take this on?

**Comment by benjaoming** [comment]
Yes, all good and well-defined, would love to fix it :+1:

**Comment by benjaoming** [comment]
Just updated the title to @humitos's point that it's not necessarily DB calls that cause the timeouts, but I'll update the issue with the cause once identified.

**Comment by humitos** [comment]
This is probably hitting the same issue we had in https://github.com/readthedocs/readthedocs.org/pull/10446 where we needed to call a raw SQL query to avoid Django performing a `SELECT` query to trigger all the `pre_/post_` signals.

**Comment by stsewd** [comment]
Just noting that this also happens when a project has lots of versions (>1K).

**Comment by agjohnson** [comment]
Noting that this also affects organizations trying to delete their own organization, which also compounds across multiple projects. In https://github.com/readthedocs/readthedocs-corporate/issues/2017 I mentioned using a task to perform deletions as well.

**Comment by stsewd** [comment]
These are all the queries that are executed when a project is deleted... We may want to check if `_raw_delete` avoids some of those queries.

[project-delete-qs.txt](https://github.com/user-attachments/files/21778483/project-delete-qs.txt)

**Comment by ericholscher** [comment]
I tried a few approaches here, but the PRs are mostly stalled. I think https://github.com/readthedocs/readthedocs.org/pull/12386 is probably the most likely thing that will actually meaningfully improve things.

**Comment by stsewd** [comment]
Not sure if https://github.com/readthedocs/readthedocs.org/pull/12456 will help for all cases, but it will make it faster for some. If that doesn't work, we could do the deletion in the background, and show a message to the user `your project/organization has been scheduled for deletion`, and do the deletion inside a task. Users will still see the project/organization for some minutes, but shouldn't time out like it does now.

**Comment by ericholscher** [comment]
Yea, I think doing it in the BG is probably the obvious next step.

**Comment by agjohnson** [comment]
Yeah, the UX of a background deletion seems like it would be fine to start with. I think users will understand a slow deletion in some cases.

We have also talked about adding an attribute on Project/Organization to signal the project is "deleted" while the async task runs, which we could at least use to visually disable or flag the project. But I feel this can come later, if at all.
